airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -197,28 +227,36 @@ def _read_incremental(
                     # Break from slice loop to save state and exit from _read_incremental function.
                     break
 
-            yield self._checkpoint_state(stream_name, stream_state, connector_state, logger)
+            yield self._checkpoint_state(stream_instance, stream_state, connector_state)
             if self._limit_reached(internal_config, total_records_counter):
                 return
 
     def _read_full_refresh(
-        self, stream_instance: Stream, configured_stream: ConfiguredAirbyteStream, internal_config: InternalConfig
+        self,
+        stream_instance: Stream,
+        configured_stream: ConfiguredAirbyteStream,
+        internal_config: InternalConfig,
     ) -> Iterator[AirbyteMessage]:
         slices = stream_instance.stream_slices(sync_mode=SyncMode.full_refresh, cursor_field=configured_stream.cursor_field)
         total_records_counter = 0
         for slice in slices:
             records = stream_instance.read_records(
-                stream_slice=slice, sync_mode=SyncMode.full_refresh, cursor_field=configured_stream.cursor_field
+                stream_slice=slice,
+                sync_mode=SyncMode.full_refresh,
+                cursor_field=configured_stream.cursor_field,
             )
             for record in records:
                 yield self._as_airbyte_record(configured_stream.stream.name, record)
                 total_records_counter += 1
                 if self._limit_reached(internal_config, total_records_counter):
                     return
 
-    def _checkpoint_state(self, stream_name, stream_state, connector_state, logger):
-        logger.info(f"Setting state of {stream_name} stream to {stream_state}")
-        connector_state[stream_name] = stream_state
+    def _checkpoint_state(self, stream, stream_state, connector_state):
```

Comment:
@keu could we maintain the invariant that the state output from this method is always the same as `stream.state`? I think the only thing I'm concerned about is that if I define both `stream.state` and `get_updated_state` then there are two potentially different states floating around which will lead to confusing behavior. 

Can we always maintain the invariant that whatever is stored in `stream.state` contains the state object being output?

Reply:
@sherifnada I'm not sure how we can achieve this, `_checkpoint_state ` will always return value from `stream.state` if there is any, if not it will fallback to the state obtained from `get_updated_state`,
so what is the problem here?

Reply:
so the contract is `IncrementalMixin` implementation always takes precedence over `get_updated_state`? sounds fine w me. 

Should we add this to the docs?

Reply:
sure, will do